core/src/Telemetry/TelemetryHelper.cs

```diff
@@ -0,0 +1,79 @@
+﻿//----------------------------------------------------------------------
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Core.Telemetry
+{
+    internal class TelemetryHelper : IDisposable
+    {
+        private readonly string _requestId;
+        private readonly EventBase _eventToEnd;
+        private readonly ITelemetry _telemetry;
+        private readonly bool _shouldFlush;
+
+        public TelemetryHelper(ITelemetry telemetry, string requestId, EventBase eventToStart, EventBase eventToEnd, bool shouldFlush)
+        {
+            _telemetry = telemetry;
+            _requestId = requestId;
+            _eventToEnd = eventToEnd;
+            _shouldFlush = shouldFlush;
+            _telemetry.StartEvent(_requestId, eventToStart);
+        }
+
+        #region IDisposable Support
+        private bool _disposedValue = false; // To detect redundant calls
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!_disposedValue)
+            {
+                if (disposing)
+                {
+                    _telemetry.StopEvent(_requestId, _eventToEnd);
+                    if (_shouldFlush)
+                    {
+                        _telemetry.Flush(_requestId);
+                    }
+                }
+
+                _disposedValue = true;
+            }
+        }
+
+        // This code added to correctly implement the disposable pattern.
+        public void Dispose()
+        {
+            Dispose(true);

```

Comment:
You can also call GC.SuppressFinalize(this); here to tell the Garbage Collector that the object does not need to be cleanup anymore because we did it manually. 

https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/dispose-pattern

Reply:
This is only needed if you implement a Finalizer, which you should only do if you are disposing unmanaged resources.  I used the VS "implement disposable pattern" and it generates a bunch of comments to that effect, and we had this same concern on a previous team.  But if you're only using managed resources (which we are in this case since we're just using IDisposable to call something on scope exit) then we don't need to suppress the finalizer since there isn't one.

Reply:
Ah, cool, I didn't realize that. Also, appologies for some duplicate comments, my Internet connection went down exactly when I was hitting submit :)